### PR TITLE
[interp] Simplify the main interpreter loop code to speed up compilation.

### PR DESCRIPTION
**Pull request by vargaz**

### CODE REVIEW

File: mono/mini/interp/interp.c [hunk 1]
```diff
@@ -215,13 +211,39 @@ static void debug_enter (InterpFrame *frame, int *tracing)
...
+#endif
+
 static void
 set_resume_state (ThreadContext *context, InterpFrame *frame)
 {
 	frame->ex = NULL;
 	context->has_resume_state = 0;
 	context->handler_frame = NULL;
 	context->handler_ei = NULL;
+	/* We have thrown an exception from a finally block. Some of the leave targets were unwinded already */ \
```
Comment (jaykrell) [inline mono/mini/interp/interp.c]: No need for trailing backslash.

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

File: mono/mini/interp/interp.c [hunk 2]
```diff
@@ -215,9 +211,35 @@ static void debug_enter (InterpFrame *frame, int *tracing)
...
+#define MINT_IN_CASE(x) case x:
+#define MINT_IN_DISPATCH(op) goto main_loop;
+#define MINT_IN_BREAK break
+#define MINT_IN_DEFAULT default:
+#endif
+
 static void
 set_resume_state (ThreadContext *context, InterpFrame *frame)
 {
+	/* We have thrown an exception from a finally block. Some of the leave targets were unwinded already */ \
```
Comment (jaykrell) [inline mono/mini/interp/interp.c]: Still no need for backslash.

File: mono/mini/interp/interp.c [hunk 3]
```diff
@@ -925,6 +942,19 @@ interp_throw (ThreadContext *context, MonoException *ex, InterpFrame *frame, gco
 
 #define THROW_EX(exception,ex_ip) THROW_EX_GENERAL ((exception), (ex_ip), FALSE)
 
+#define THROW_EX_OVF(ip) do { \
```
Comment (jaykrell) [inline mono/mini/interp/interp.c]: do/while probably not needed here, assuming `THROW_EX` is well written.

File: mono/mini/interp/interp.c [hunk 4]
```diff
@@ -925,6 +942,19 @@ interp_throw (ThreadContext *context, MonoException *ex, InterpFrame *frame, gco
...
+	THROW_EX (mono_get_exception_overflow (), ip); \
+	} while (0)
+
+#define THROW_EX_DIV_ZERO(ip) do { \
+	THROW_EX (mono_get_exception_divide_by_zero (), ip);	\
+	} while (0)
```
Comment (jaykrell) [inline mono/mini/interp/interp.c]: do/while probably not needed here, assuming `THROW_EX` is well written.

File: mono/mini/interp/interp-internals.h [hunk 5]
```diff
@@ -147,6 +147,8 @@ struct _InterpFrame {
 	/* exception info */
 	const unsigned short  *ip;
 	MonoException     *ex;
+	GSList *finally_ips;
+	const unsigned short *endfinally_ip;
```
Comment (jaykrell) [inline mono/mini/interp/interp-internals.h]: Not good to put stuff in address-taken structs.
I have a fix coming.

**Pull request merged**